File: src/lib/cli/commands/completion.mjs
import { parseArgs } from 'node:util';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';

import { EXIT, LegionError } from '../../errors.mjs';
import { loadCanonicalHostKeyRing, loadHostKeyRing } from '../../guard/compat/host/keys.mjs';
import { digestValue, isDigest } from '../../contracts/arcane/canonical.mjs';
import { SessionBindingStore } from '../../host/arcane/session-binding.mjs';
import { ContractSealStore } from './governance/delivery/contract-seal-store.mjs';
import { HostEventLedger } from '../../host/arcane/host-event-ledger.mjs';
import { AuthorityInvocationProofIssuer } from '../../contracts/arcane/authority-invocation-proof.mjs';
import { PendingTerminalOperationStore } from '../../verification/arcane/pending-terminal-operation-store.mjs';
import { requireCanonicalAdvisoryProfile } from '../../verification/arcane/advisory-profile.mjs';
import { ReceiptStore } from '../../guard/compat/audit/receipt-store.mjs';
import { sealEvidence } from '../../verification/arcane/evidence-envelope.mjs';
import { signRecord, EVIDENCE_RECEIPT_BOUND_FIELDS } from '../../guard/compat/audit/receipt-auth.mjs';
import { completionIntegratedStateForRepositories, latestScopedMaterialChange } from '../../verification/arcane/completion-state.mjs';
import { DependencyLedger } from '../../verification/arcane/invalidation.mjs';

const session = (value, env) => value || env.CODEX_THREAD_ID || env.CLAUDE_CODE_SESSION_ID || env.CLAUDE_SESSION_ID || env.CODEX_SESSION_ID || null;

function advisoryClaimFor(outcome, seal) {
  const requested = outcome?.advisoryClaim;
  if (requested == null) return null;
  if (!requested || typeof requested !== 'object' || Object.keys(requested).sort().join(',') !== 'artifactDigest,briefDigest' || !isDigest(requested.artifactDigest) || !isDigest(requested.briefDigest)) {
    throw new LegionError('ARC_SCHEMA_INVALID: advisory claim requires exact artifactDigest & briefDigest', { code: 'ARC_SCHEMA_INVALID', exitCode: EXIT.USAGE });
  }
  const profile = requireCanonicalAdvisoryProfile(seal.contract?.advisoryProfile);
  return {
    required: true,
    artifactDigest: requested.artifactDigest,
    briefDigest: requested.briefDigest,
    bundleId: profile.bundleId,
    bundleVersion: profile.bundleVersion,
    profileId: profile.profileId,
    manifestDigest: profile.manifestDigest,
    profileDigest: profile.profileDigest,
  };
}
function highRiskContextFor(outcome) {
  const context = outcome?.highRiskContext;
  if (context == null) return null;
  const fields = ['blastRadius', 'intentRestatement', 'whySafe'];
  if (!context || typeof context !== 'object' || Object.keys(context).sort().join(',') !== fields.sort().join(',') || fields.some((field) => typeof context[field] !== 'string' || !context[field].trim())) {
    throw new LegionError('ARC_SCHEMA_INVALID: highRiskContext requires exact non-empty blastRadius, intentRestatement, & whySafe', { code: 'ARC_SCHEMA_INVALID', exitCode: EXIT.USAGE });
  }
  return Object.fromEntries(fields.map((field) => [field, context[field]]));
}
function completionRepositories(cwd, binding, scope) {
  const roots = binding.delivery?.repositories?.map((repo) => repo.root).filter(Boolean);
  return (roots?.length ? roots : [cwd]).map((root) => ({ cwd: root, scope }));
}
function latestCompletionChange(receiptStore, runId, repositories) {
  return repositories.map(({ cwd, scope }) => latestScopedMaterialChange(receiptStore, runId, scope, cwd)).filter(Boolean).sort().at(-1) ?? null;
}

export async function runCompletion(argv, { stdout, env, cwd }) {
  const [sub, ...rest] = argv;
  if (sub === '--help' || sub === 'help') {
    stdout.write('Usage: legion completion claim|evidence --file <outcome.json> [--session <id>]\n');
    return { exitCode: EXIT.PASS };
  }
  if (sub === 'evidence') return runEvidence(rest, { stdout, env, cwd });
  if (sub !== 'claim') throw new LegionError('completion requires claim', { code: 'USAGE', exitCode: EXIT.USAGE });
  const values = parseArgs({ args: rest, strict: true, options: { file: { type: 'string' }, session: { type: 'string' }, 'key-dir': { type: 'string' } } }).values;
  const sessionId = session(values.session, env);
  if (!values.file || !sessionId) throw new LegionError('completion claim requires --file & known session', { code: 'USAGE', exitCode: EXIT.USAGE });

  const keys = values['key-dir'] || env.ARCANE_KEY_DIR ? loadHostKeyRing({ dir: values['key-dir'] || env.ARCANE_KEY_DIR }) : loadCanonicalHostKeyRing();
  const binding = new SessionBindingStore({ root: join(cwd, '.audit', 'arcane', 'session-bindings') }).getBinding(sessionId);
  if (!binding?.contractId || !binding.taskId) throw new LegionError('ARC_NO_CONTRACT', { code: 'ARC_NO_CONTRACT', exitCode: EXIT.INCOMPLETE });
  const seal = new ContractSealStore({ root: join(cwd, '.audit', 'arcane', 'contract-seals') }).get(binding.contractId, binding.contractVersion);
  if (!seal || seal.contractDigest !== binding.contractDigest) throw new LegionError('ARC_CONTRACT_VERSION_MISMATCH', { code: 'ARC_CONTRACT_VERSION_MISMATCH', exitCode: EXIT.INCOMPLETE });

  const ledger = new HostEventLedger({ root: join(cwd, '.audit', 'arcane', 'host-events'), keyRing: keys, keyId: keys.activeKeyId() });
  if (!ledger.verify().allowed) throw new LegionError('ARC_STORE_CORRUPT: legion host events inspect', { code: 'ARC_STORE_CORRUPT', exitCode: EXIT.INCOMPLETE });
  const event = ledger.records().reverse().find((record) => record.sessionId === sessionId);
  if (!event || !['legion', 'alchemist'].includes(event.observedAuthority)) throw new LegionError('ARC_AUTHORITY_NOT_ASSERTED: current Legion or assigned Alchemist event required', { code: 'ARC_AUTHORITY_NOT_ASSERTED', exitCode: EXIT.INCOMPLETE });

  const rootKeyId = keys.activeKeyId();
  const issuer = new AuthorityInvocationProofIssuer({ root: join(cwd, '.audit', 'arcane', 'authority-invocations'), keyRing: keys, keyId: rootKeyId, ledgerStore: ledger });
  const proof = issuer.issue({ ledger: event, binding: { ...binding, sessionId }, purpose: 'completion-claim', role: event.observedAuthority }).proof;
  const store = new PendingTerminalOperationStore({ root: join(cwd, '.audit', 'arcane', 'terminal-operations'), keyRing: keys, keyId: keys.activeKeyId() });
  const outcome = JSON.parse(readFileSync(values.file, 'utf8'));
  const claim = store.mint({
    invocationProofDigest: digestValue(proof),
    producerAuthority: event.observedAuthority,
    runId: binding.runId,
    taskId: binding.taskId,
    contractId: binding.contractId,
    contractVersion: binding.contractVersion,
    contractDigest: binding.contractDigest,
    sourceRevision: seal.sourceRevision,
    turnCorrelationDigest: event.turnCorrelationDigest,
    expectedStopOrdinal: (event.stopOrdinal || 0) + 1,
    outcomeSummaryDigest: digestValue(outcome.outcomeSummary || ''),
    artifactStateDigest: digestValue(outcome.artifactState || ''),
    advisoryClaim: advisoryClaimFor(outcome, seal),
    highRiskContext: highRiskContextFor(outcome),
  });
  const consumed = issuer.consume(proof, { artifactDigest: digestValue(claim) });
  if (!consumed.allowed) throw new LegionError(consumed.code, { code: consumed.code, exitCode: EXIT.INCOMPLETE });
  const stored = store.append(claim);
  if (!stored.allowed) throw new LegionError(stored.code, { code: stored.code, exitCode: EXIT.INCOMPLETE });
  stdout.write(`${JSON.stringify({ claimId: stored.detail.claimId, invocationProofDigest: digestValue(proof) })}\n`);
  return { exitCode: EXIT.PASS };
}

function runEvidence(argv, { stdout, env, cwd }) {
  const values = parseArgs({ args: argv, strict: true, options: { file: { type: 'string' }, session: { type: 'string' }, 'key-dir': { type: 'string' } } }).values;
  const sessionId = session(values.session, env); if (!values.file || !sessionId) throw new LegionError('completion evidence requires --file & known session', { code: 'USAGE', exitCode: EXIT.USAGE });
  const keys = values['key-dir'] || env.ARCANE_KEY_DIR ? loadHostKeyRing({ dir: values['key-dir'] || env.ARCANE_KEY_DIR }) : loadCanonicalHostKeyRing();
  const binding = new SessionBindingStore({ root: join(cwd, '.audit', 'arcane', 'session-bindings') }).getBinding(sessionId);
  const seal = binding?.contractId ? new ContractSealStore({ root: join(cwd, '.audit', 'arcane', 'contract-seals') }).get(binding.contractId, binding.contractVersion) : null;
  if (!binding?.taskId || !seal || seal.contractDigest !== binding.contractDigest) throw new LegionError('ARC_NO_CONTRACT', { code: 'ARC_NO_CONTRACT', exitCode: EXIT.INCOMPLETE });
  const ledger = new HostEventLedger({ root: join(cwd, '.audit', 'arcane', 'host-events'), keyRing: keys, keyId: keys.activeKeyId() });
  const event = ledger.records().reverse().find((record) => record.sessionId === sessionId);
  if (!ledger.verify().allowed || !event || event.observedAuthority !== 'oracle') throw new LegionError('ARC_AUTHORITY_NOT_ASSERTED: current Oracle host event required', { code: 'ARC_AUTHORITY_NOT_ASSERTED', exitCode: EXIT.INCOMPLETE });
  const rootKeyId = keys.activeKeyId();
  const issuer = new AuthorityInvocationProofIssuer({ root: join(cwd, '.audit', 'arcane', 'authority-invocations'), keyRing: keys, keyId: rootKeyId, ledgerStore: ledger });
  const authorityProof = issuer.issue({ ledger: event, binding: { ...binding, sessionId }, purpose: 'completion-claim', role: 'oracle' }).proof;
  const report = JSON.parse(readFileSync(values.file, 'utf8')); const item = report?.evidence;
  const criterion = seal.contract.acceptanceCriteria.find((entry) => entry.id === item?.acceptanceId);
  if (!criterion || !item || !['requirementId','productionSymbol','liveConsumer','acceptanceSurface'].every((key) => typeof item[key] === 'string' && item[key])) throw new LegionError('ARC_SCHEMA_INVALID: evidence must map sealed acceptance to requirement, symbol, consumer, & surface', { code: 'ARC_SCHEMA_INVALID', exitCode: EXIT.USAGE });
  const receipts = new ReceiptStore({ root: join(cwd, '.audit', 'arcane', 'receipts') });
  const repositories = completionRepositories(cwd, binding, seal.contract.scope.own);
  const latest = latestCompletionChange(receipts, binding.runId, repositories);
  const observation = { acceptanceId: criterion.id, requirementId: item.requirementId, productionSymbol: item.productionSymbol, liveConsumer: item.liveConsumer, acceptanceSurface: item.acceptanceSurface, integratedState: completionIntegratedStateForRepositories(repositories), latestMaterialChange: latest, contractVersion: binding.contractVersion, contractDigest: binding.contractDigest, sourceRevision: seal.sourceRevision, validUntil: item.validUntil, authorityProofDigest: digestValue(authorityProof) };
  const dependencies = [
    { dimension: 'source-digest', ref: item.liveConsumer, digest: digestValue(seal.sourceRevision) },
    { dimension: 'upstream-contract-revision', ref: binding.contractId, digest: binding.contractDigest },
  ];
  const { receipt } = sealEvidence({ runId: binding.runId, taskId: binding.taskId, contractId: binding.contractId, producerAuthority: 'oracle', capability: 'oracle-acceptance-observation', observation, evidenceClass: 'deterministic', sourceRevision: seal.sourceRevision, dependencies, authentication: { issuerIdentity: 'oracle', verificationMethod: 'capability-signature', perMessage: true, verifiedAt: new Date().toISOString() }, replayDefense: { nonce: authorityProof.invocationId, sequence: 1, freshnessWindowSeconds: 3600, freshAt: new Date().toISOString() }, observedAt: new Date().toISOString() });
  receipt.authentication = { ...receipt.authentication, ...signRecord(receipt, { keyRing: keys, keyId: rootKeyId, boundFields: EVIDENCE_RECEIPT_BOUND_FIELDS }) };
  const consumed = issuer.consume(authorityProof, { artifactDigest: digestValue(receipt) });
  if (!consumed.allowed) throw new LegionError(consumed.code, { code: consumed.code, exitCode: EXIT.INCOMPLETE });
  receipts.append(receipt);
  const dependenciesLedger = new DependencyLedger({ root: join(cwd, '.audit', 'arcane', 'dependency-ledger') });
  dependenciesLedger.register(receipt.evidenceId, dependencies);
  dependenciesLedger.link(receipt.evidenceId, { criterionId: criterion.id });
  stdout.write(`${JSON.stringify({ evidenceId: receipt.evidenceId, acceptanceId: criterion.id })}\n`); return { exitCode: EXIT.PASS };
}

File: engine/bins/legion/src/commands/completion.rs
//! Authenticated completion claim and Oracle evidence persistence.
use super::{CommandError, CommandResult};
use crate::cli::CommonArgs;
use legion_arcane::receipt_auth::{sign_record, verify_record};
use legion_arcane::{AuthorityInvocationProofIssuer, KeyRing, ReceiptStore, SessionBindingStore};
use legion_contracts::canonical_digest;
use serde_json::{json, Map, Value};
use sha2::{Digest, Sha256};
use std::{
    fs,
    path::{Path, PathBuf},
    time::{SystemTime, UNIX_EPOCH},
};

const HOST_FIELDS: [&str; 19] = [
    "schemaVersion",
    "kind",
    "eventId",
    "eventSequence",
    "previousDigest",
    "turnCorrelationDigest",
    "stopOrdinal",
    "adapter",
    "eventType",
    "sessionId",
    "runId",
    "taskId",
    "contractId",
    "contractVersion",
    "contractDigest",
    "sourceRevision",
    "observedAuthority",
    "payloadDigest",
    "observedAt",
];
const EVIDENCE_FIELDS: [&str; 13] = [
    "schemaVersion",
    "kind",
    "evidenceId",
    "runId",
    "taskId",
    "contractId",
    "producerAuthority",
    "capability",
    "observation",
    "evidenceClass",
    "sourceRevision",
    "dependsOn",
    "observedAt",
];
fn arg(a: &[String], n: &str) -> Option<String> {
    a.windows(2).find(|p| p[0] == n).map(|p| p[1].clone())
}
fn session(a: &[String]) -> Result<String, CommandError> {
    arg(a, "--session")
        .filter(|v| !v.is_empty())
        .or_else(|| std::env::var("CODEX_THREAD_ID").ok())
        .or_else(|| std::env::var("CLAUDE_CODE_SESSION_ID").ok())
        .or_else(|| std::env::var("CLAUDE_SESSION_ID").ok())
        .or_else(|| std::env::var("CODEX_SESSION_ID").ok())
        .filter(|v| !v.is_empty())
        .ok_or_else(|| {
            CommandError::usage("completion requires --session <id> for authenticated binding")
        })
}
fn root() -> Result<PathBuf, CommandError> {
    std::env::current_dir().map_err(super::io_error)
}
fn ring(a: &[String]) -> Result<KeyRing, CommandError> {
    let p = arg(a, "--key-dir")
        .map(PathBuf::from)
        .or_else(|| std::env::var_os("ARCANE_KEY_DIR").map(PathBuf::from));
    match p {
        Some(p) => KeyRing::load_dir(&p).map_err(|e| CommandError::incomplete(e.to_string())),
        None => KeyRing::load_canonical().map_err(|e| CommandError::incomplete(e.to_string())),
    }
}
fn hash(s: &str) -> String {
    let mut h = Sha256::new();
    h.update(s.as_bytes());
    hex::encode(h.finalize())
}
fn seal(root: &Path, b: &Value) -> Result<Value, CommandError> {
    let p = legion_arcane::state_paths::state_file(
        &root.join(".audit/arcane/contract-seals"),
        "arcane.contract-seal.key.v1",
        &[
            b["contractId"].as_str().unwrap_or_default().into(),
            b["contractVersion"]
                .as_u64()
                .unwrap_or_default()
                .to_string(),
        ],
    )
    .map_err(|e| CommandError::integrity(e.to_string()))?;
    serde_json::from_slice(
        &fs::read(p).map_err(|_| CommandError::incomplete("ARC_CONTRACT_VERSION_MISMATCH"))?,
    )
    .map_err(|_| CommandError::incomplete("ARC_STORE_CORRUPT: unreadable contract seal"))
}
fn events(root: &Path, session: &str, k: &KeyRing) -> Result<Vec<Value>, CommandError> {
    let dir = root.join(".audit/arcane/host-events");
    let mut files = fs::read_dir(&dir)
        .map_err(|_| {
            CommandError::incomplete(
                "ARC_STORE_CORRUPT: host event ledger continuity is unavailable",
            )
        })?
        .flatten()
        .filter(|e| {
            e.file_name().to_string_lossy().ends_with(".json")
                && e.file_name().to_string_lossy().len() == 21
        })
        .collect::<Vec<_>>();
    files.sort_by_key(|e| e.file_name());
    let mut out = Vec::new();
    let mut prev: Option<String> = None;
    for f in files {
        let v: Value = serde_json::from_slice(
            &fs::read(f.path()).map_err(|_| CommandError::incomplete("ARC_STORE_CORRUPT"))?,
        )
        .map_err(|_| CommandError::incomplete("ARC_STORE_CORRUPT"))?;
        let auth = v
            .get("authentication")
            .ok_or_else(|| CommandError::incomplete("ARC_STORE_CORRUPT"))?;
        let d = verify_record(&v, Some(auth), k, &HOST_FIELDS, &Map::new(), None)
            .map_err(|_| CommandError::incomplete("ARC_STORE_CORRUPT"))?;
        if !d.allowed || v["previousDigest"].as_str() != prev.as_deref() {
            return Err(CommandError::incomplete(
                "ARC_STORE_CORRUPT: host event ledger continuity is unavailable",
            ));
        }
        prev = Some(canonical_digest(&v).map_err(|e| CommandError::integrity(e.to_string()))?);
        if v["sessionId"].as_str() == Some(session) {
            out.push(v)
        }
    }
    if out.is_empty() {
        return Err(CommandError::incomplete(
            "ARC_AUTHORITY_NOT_ASSERTED: current authenticated authority event required",
        ));
    }
    Ok(out)
}
fn claim(root: &Path, a: &[String], file: &str, b: &Value, s: &str, k: KeyRing) -> CommandResult {
    let outcome: Value = serde_json::from_slice(
        &fs::read(file)
            .map_err(|e| CommandError::usage(format!("invalid completion artifact: {e}")))?,
    )
    .map_err(|e| CommandError::usage(format!("invalid completion artifact: {e}")))?;
    if !outcome.is_object() {
        return Err(CommandError::usage(
            "completion artifact must contain a JSON object",
        ));
    }
    if outcome.get("outcomeSummary").is_none() || outcome.get("artifactState").is_none() {
        return Err(CommandError::usage(
            "completion claim artifact requires outcomeSummary and artifactState",
        ));
    }
    let es = events(root, s, &k)?;
    let event = es
        .iter()
        .rev()
        .find(|e| {
            matches!(
                e["observedAuthority"].as_str(),
                Some("legion" | "alchemist")
            )
        })
        .ok_or_else(|| {
            CommandError::incomplete(
                "ARC_AUTHORITY_NOT_ASSERTED: current Legion or assigned Alchemist event required",
            )
        })?;
    let issuer = AuthorityInvocationProofIssuer::new(
        root.join(".audit/arcane/authority-invocations"),
        k.clone(),
        k.active_key_id()
            .map_err(|e| CommandError::incomplete(e.to_string()))?,
    );
    let proof = issuer
        .issue(
            event,
            b,
            "completion-claim",
            event["observedAuthority"].as_str().unwrap(),
        )
        .map_err(|e| CommandError::incomplete(e))?;
    let claim = json!({"schemaVersion":1,"kind":"arcane-terminal-operation-claim","claimId":format!("claim_{}",hash(&format!("{}:{}",b["runId"],canonical_digest(&outcome).unwrap_or_default()))),"invocationProofDigest":canonical_digest(&proof).unwrap_or_default(),"producerAuthority":event["observedAuthority"],"runId":b["runId"],"taskId":b["taskId"],"contractId":b["contractId"],"contractVersion":b["contractVersion"],"contractDigest":b["contractDigest"],"sourceRevision":seal(root,b)?["sourceRevision"],"outcomeSummaryDigest":canonical_digest(outcome.get("outcomeSummary").unwrap()).unwrap_or_default(),"artifactStateDigest":canonical_digest(outcome.get("artifactState").unwrap()).unwrap_or_default(),"turnCorrelationDigest":event["turnCorrelationDigest"],"expectedStopOrdinal":event["stopOrdinal"].as_u64().unwrap_or_default()+1});
    let dir = root.join(".audit/arcane/terminal-operations");
    fs::create_dir_all(&dir).map_err(super::io_error)?;
    let path = dir.join(format!(
        "{}.json",
        claim["claimId"].as_str().unwrap_or_default()
    ));
    if path.exists() {
        return Err(CommandError::incomplete("ARC_REPLAY_NONCE_SEEN"));
    }
    issuer
        .consume(&proof, Some(&canonical_digest(&claim).unwrap_or_default()))
        .map_err(|e| CommandError::incomplete(e))?;
    fs::write(path, serde_json::to_vec(&claim).map_err(super::io_error)?)
        .map_err(super::io_error)?;
    Ok(json!({"claimId":claim["claimId"],"invocationProofDigest":claim["invocationProofDigest"]}))
}
fn evidence(
    root: &Path,
    a: &[String],
    file: &str,
    b: &Value,
    s: &str,
    k: KeyRing,
) -> CommandResult {
    let report: Value = serde_json::from_slice(
        &fs::read(file)
            .map_err(|e| CommandError::usage(format!("invalid completion artifact: {e}")))?,
    )
    .map_err(|e| CommandError::usage(format!("invalid completion artifact: {e}")))?;
    let item = report.get("evidence").ok_or_else(|| {
        CommandError::usage("completion evidence artifact requires evidence object")
    })?;
    let se = seal(root, b)?;
    let criteria = se["contract"]["acceptanceCriteria"]
        .as_array()
        .cloned()
        .unwrap_or_default();
    let aid=item["acceptanceId"].as_str().ok_or_else(||CommandError::usage("ARC_SCHEMA_INVALID: evidence must map sealed acceptance to requirement, symbol, consumer, & surface"))?;
    if !criteria.iter().any(|x| x["id"].as_str() == Some(aid)) {
        return Err(CommandError::usage(
            "ARC_SCHEMA_INVALID: evidence acceptanceId is not sealed",
        ));
    }
    for key in [
        "requirementId",
        "productionSymbol",
        "liveConsumer",
        "acceptanceSurface",
    ] {
        if item[key].as_str().is_none_or(|x| x.is_empty()) {
            return Err(CommandError::usage("ARC_SCHEMA_INVALID: evidence must map sealed acceptance to requirement, symbol, consumer, & surface"));
        }
    }
    let es = events(root, s, &k)?;
    let event = es
        .iter()
        .rev()
        .find(|e| e["observedAuthority"] == "oracle")
        .ok_or_else(|| {
            CommandError::incomplete(
                "ARC_AUTHORITY_NOT_ASSERTED: current Oracle host event required",
            )
        })?;
    let issuer = AuthorityInvocationProofIssuer::new(
        root.join(".audit/arcane/authority-invocations"),
        k.clone(),
        k.active_key_id()
            .map_err(|e| CommandError::incomplete(e.to_string()))?,
    );
    let proof = issuer
        .issue(event, b, "completion-claim", "oracle")
        .map_err(|e| CommandError::incomplete(e))?;
    let obs = json!({"acceptanceId":aid,"requirementId":item["requirementId"],"productionSymbol":item["productionSymbol"],"liveConsumer":item["liveConsumer"],"acceptanceSurface":item["acceptanceSurface"],"contractVersion":b["contractVersion"],"contractDigest":b["contractDigest"],"sourceRevision":se["sourceRevision"],"authorityProofDigest":canonical_digest(&proof).unwrap_or_default(),"validUntil":item["validUntil"]});
    let deps = json!([]);
    let mut rec = json!({"schemaVersion":1,"kind":"legion-evidence-capability-receipt","evidenceId":format!("evidence_{}",hash(&format!("{}:{}",b["runId"],aid))),"runId":b["runId"],"taskId":b["taskId"],"contractId":b["contractId"],"producerAuthority":"oracle","capability":"oracle-acceptance-observation","observation":obs,"evidenceClass":"deterministic","sourceRevision":se["sourceRevision"],"dependsOn":deps,"observedAt":now()});
    rec["authentication"] = sign_record(
        &rec,
        &k,
        &k.active_key_id()
            .map_err(|e| CommandError::incomplete(e.to_string()))?,
        &EVIDENCE_FIELDS,
        None,
    )
    .map_err(|e| CommandError::incomplete(e.to_string()))?;
    let rs =
        ReceiptStore::new(root.join(".audit/arcane/receipts")).map_err(CommandError::incomplete)?;
    let digest = canonical_digest(&rec).unwrap_or_default();
    issuer
        .consume(&proof, Some(&digest))
        .map_err(|e| CommandError::incomplete(e))?;
    rs.append(&rec);
    Ok(json!({"evidenceId":rec["evidenceId"],"acceptanceId":aid}))
}
pub fn run(args: CommonArgs) -> CommandResult {
    let a = args
        .args
        .iter()
        .map(|v| v.to_string_lossy().into_owned())
        .collect::<Vec<_>>();
    let sub = a.first().map(String::as_str).ok_or_else(|| {
        CommandError::usage(
            "completion requires claim|evidence --file <outcome.json> [--session <id>]",
        )
    })?;
    if sub == "--help" || sub == "help" {
        return Ok(
            json!({"__raw":"Usage: legion completion claim|evidence --file <outcome.json> [--session <id>]\n"}),
        );
    }
    if !matches!(sub, "claim" | "evidence") {
        return Err(CommandError::usage("completion requires claim"));
    }
    let f = arg(&a, "--file").ok_or_else(|| {
        CommandError::usage(format!("completion {sub} requires --file <outcome.json>"))
    })?;
    let s = session(&a)?;
    let root = root()?;
    let k = ring(&a)?;
    let bs = SessionBindingStore::new(root.join(".audit/arcane/session-bindings"));
    let b = bs.get(&s).ok_or_else(|| {
        CommandError::incomplete("completion requires an authenticated session binding")
    })?;
    if b["contractId"].is_null() || b["taskId"].is_null() {
        return Err(CommandError::incomplete("ARC_NO_CONTRACT"));
    }
    if sub == "claim" {
        claim(&root, &a, &f, &b, &s, k)
    } else {
        evidence(&root, &a, &f, &b, &s, k)
    }
}
fn now() -> String {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs()
        .to_string()
}
